### ISSUE TITLE
"adapter " option in udp out

### ISSUE THREAD

**Issue by vasudevavarun** [Haivision/srt#956]
Hi,
SRT Version: 1.4.0
OS: Ubuntu 18.04 LTS

I'm trying to send the udp multicast output on a particular adapter using the following syntax.
**srt-live-transmit "srt://xxx.xxx.xxx.xxx:xxxx?mode=client&adapter=xxx.xxx.xxx.xxx" "udp://224.1.1.1:4111?adapter=172.17.0.1"**

But no matter what adapter I mention the output traffic still goes on a random ethernet,
Can you please tell if the adapter option is available for udp output or not.

Thanks

**Comment by ethouris** [comment]
It is. Please use `-v` option so that this information can be displayed.

**Comment by ethouris** [comment]
For multicast, the UDP medium uses two extra attributes: `adapter` and `source`. The latter would use `IP_ADD_SOURCE_MEMBERSHIP` instead of `IP_ADD_MEMBERSHIP`.

**Comment by vasudevavarun** [comment]
Hello,
Thank you for the prompt reply.
While using '-v' I got the following output. I am able to see the output binding to 172.17.0.1 but the multicast traffic is still going on the other interface.

**srt-live-transmit -v "srt://x.x.x.x:7001?mode=client&adapter=192.168.27.5" "udp://224.1.1.1:4111?adapter=172.17.0.1"**

Media path: 'srt://x.x.x.x:7001?mode=client&adapter=192.168.27.5' --> 'udp://224.1.1.1:4111?adapter=172.17.0.1'
SRT parameters specified:

	adapter = '192.168.27.5'
	mode = 'client'
Opening SRT source caller on x.x.x.x:7001
Connecting to x.x.x.x:7001
Multicast: home address: **172.17.0.1:4111**
Multicast(POSIX): will bind to IGMP address: 224.1.1.1
SRT source connected



> For multicast, the UDP medium uses two extra attributes: `adapter` and `source`. The latter would use `IP_ADD_SOURCE_MEMBERSHIP` instead of `IP_ADD_MEMBERSHIP`.

I'm not able to understand the adapter and source co-relation, If you can give an example/syntax that would be a great help.

Thanks.

**Comment by ethouris** [comment]
> While using '-v' I got the following output. I am able to see the output binding to 172.17.0.1 but the multicast traffic is still going on the other interface.

I think I know where the problem is. According to the description in the manpage, setting the `ip_mreq::imr_interface` field to appropriate IP address matters for reading from the multicast address, but not for sending. Sending still undergoes the standard rules of the system, where the network interface for sending is selected basing on the route table. There are only two ways to override it: using cmsg and IP_PKTINFO, or binding the socket to the specified address.

If you can please test if adding a fix after this line:
```
            int res = setsockopt(m_sock, IPPROTO_IP, opt_name, mreq_arg, mreq_arg_size);
```
Should help:
```
::bind(m_sock, (sockaddr*)&maddr, sizeof maddr);
```

Of course, the real fix would have to be different because this actually messes things up a bit, but for your case it should do the job.

> > For multicast, the UDP medium uses two extra attributes: `adapter` and `source`. The latter would use `IP_ADD_SOURCE_MEMBERSHIP` instead of `IP_ADD_MEMBERSHIP`.
> 
> I'm not able to understand the adapter and source co-relation, If you can give an example/syntax that would be a great help.

I personally never used it so I don't know anything. You can read about that in Linux the manpage `ip(7)`.

**Comment by ethouris** [comment]
@vasudevavarun : please clear things up as to whether the above solution fixes the problem. If so, I'll prepare a PR that will provide a real fix.

**Comment by vasudevavarun** [comment]
> @vasudevavarun : please clear things up as to whether the above solution fixes the problem. If so, I'll prepare a PR that will provide a real fix.

Hi,
I'm afraid that the above option didn't work. No matter that adapter option I mention the traffic will still depend on the route table.

**Comment by vasudevavarun** [comment]
> @vasudevavarun : please clear things up as to whether the above solution fixes the problem. If so, I'll prepare a PR that will provide a real fix.

I'm sorry I was not available give me few hours I will revert very soon.

**Comment by vasudevavarun** [comment]
> @vasudevavarun : please clear things up as to whether the above solution fixes the problem. If so, I'll prepare a PR that will provide a real fix.

Hi @ethouris,
I'm not able to compile the code with the mentioned changes. can you provide a fix for it please?

**Comment by ethouris** [comment]
It compiles well on my side. You need to patch the file `apps/transmitmedia.cpp`. Here is the patch against the latest `master` - you can apply it in the repository's root directory using `patch -p1`:
```
diff --git a/apps/transmitmedia.cpp b/apps/transmitmedia.cpp
index 64cae28..d82c6eb 100644
--- a/apps/transmitmedia.cpp
+++ b/apps/transmitmedia.cpp
@@ -1047,6 +1047,7 @@ protected:
             Verb() << "Multicast(POSIX): will bind to IGMP address: " << host;
 #endif
             int res = setsockopt(m_sock, IPPROTO_IP, opt_name, mreq_arg, mreq_arg_size);
+            ::bind(m_sock, (sockaddr*)&maddr, sizeof maddr);
 
             if ( res == status_error )
             {
```

**Comment by vasudevavarun** [comment]
> It compiles well on my side. You need to patch the file `apps/transmitmedia.cpp`. Here is the patch against the latest `master` - you can apply it in the repository's root directory using `patch -p1`:
> 
> ```
> diff --git a/apps/transmitmedia.cpp b/apps/transmitmedia.cpp
> index 64cae28..d82c6eb 100644
> --- a/apps/transmitmedia.cpp
> +++ b/apps/transmitmedia.cpp
> @@ -1047,6 +1047,7 @@ protected:
>              Verb() << "Multicast(POSIX): will bind to IGMP address: " << host;
>  #endif
>              int res = setsockopt(m_sock, IPPROTO_IP, opt_name, mreq_arg, mreq_arg_size);
> +            ::bind(m_sock, (sockaddr*)&maddr, sizeof maddr);
>  
>              if ( res == status_error )
>              {
> ```

Sorry, I do not have much hands on c++.
I'm glad to say that the above patch worked very well.

**Comment by ethouris** [comment]
Alright, I'll make the fix then.

**Comment by ethouris** [comment]
@vasudevavarun Would be able to also test if the solution provided in #970 on [my PR branch](https://github.com/ethouris/srt/tree/dev-fix-multicast-sending) also fixes the problem? This time I used a little different method, more portable, but it should work.

**Comment by vasudevavarun** [comment]
> @vasudevavarun Would be able to also test if the solution provided in #970 on [my PR branch](https://github.com/ethouris/srt/tree/dev-fix-multicast-sending) also fixes the problem? This time I used a little different method, more portable, but it should work.

@ethouris - I've testing the mentioned branch. It worked as expected.

**Comment by msanis** [comment]
Hi

I am facing similar issue on a Windows 10 PC with SRT 1.4.1 using below command to receive the stream

**C:\Users\Admin\Desktop\Release>srt-live-transmit "srt://:8001?mode=listener&latency=2000" "udp://225.1.2.7:2000?adapter=10.150.181.150&ttl=64" -s:500 -v**

This PC has NIC 1 = 10.150.180.150 (for Internet Traffic) and NIC 2 = 10.150.181.150 (for multicast output)

**SRT Output:**

Media path: 'srt://:8001?mode=listener&latency=2000' --> 'udp://225.1.2.7:2000?adapter=10.150.181.150&ttl=64'
Parameters:
        latency = '2000'
        mode = 'listener'
Opening SRT source listener on :8001
Binding a server on :8001 ...
 listen...
Multicast: home address: 10.150.181.150:2000
Multicast(Windows): will bind to home address
 accept...
 connected.
Accepted SRT source connection
======= SRT STATS: sid=340732061
PACKETS     SENT:           0  RECEIVED:          1358
LOST PKT    SENT:           0  RECEIVED:             3
REXMIT      SENT:           0  RECEIVED:             8
DROP PKT    SENT:           0  RECEIVED:             0
RATE     SENDING:           0  RECEIVING:       4.4555
BELATED RECEIVED:           5  AVG TIME:    1.0891e+16
REORDER DISTANCE:           0
WINDOW      FLOW:        8192  CONGESTION:        8192  FLIGHT:           0
LINK         RTT:     146.2ms  BANDWIDTH:  146.076Mb/s
BUFFERLEFT:  SND:    12288000  RCV:           11013000

I can able to receive the output multicast and lock in TS reader on same PC. However no physical traffic is leaving the PC so other devices connected to NIC 2 can able to receive the stream. I checked the NIC ports stats and see only incoming internet traffic on NIC and aboslutely Zero traffic on 2nd interface.
![SRT](https://user-images.githubusercontent.com/30479525/93725252-7f2c3280-fba5-11ea-8301-0b2f2f0aabda.JPG)

I have a feeling the multicast output is on loopback interface and this is the reason only local TS reader can able to lock the multicast.

I tried compiling [https://github.com/ethouris/srt/tree/dev-fix-multicast-sending](url) But still the behaviour is same.
Any suggestions what causing this issue ?
